File: Test_Function_file/Test_Function.py
import numpy as np
from math import pi, exp, sqrt, cos, sin, e

class func_2D:
    def __init__(self):
        self.bound_Ackley = [(-5,5), (-5,5)]
        self.bound_Levi_N13 = [(-10,10), (-10,10)]
        self.bound_Eggholder = [(-512,512), (-512,512)]
        self.bound_Holder_table = [(-10,10), (-10,10)]
        self.bound_Rastrigin = [(-5.12,5.12), (-5.12,5.12)]
        self.bound_Schwefel = [(-500,500), (-500,500)] 
        self.n_func = self.all_func().shape[0]

    def Ackley(self, x):
        return -20.0 * exp(-0.2*sqrt(0.5*(x[0]**2 + x[1]**2))) - exp(0.5*(cos(2*pi*x[0]) + cos(2*pi*x[1]))) + e + 20

    def Levi_N13(self, x):
        return (sin(3*pi*x[0]))**2 + ((x[0] - 1)**2)*(1 + (sin(3*pi*x[1]))**2) + ((x[1] - 1)**2)*(1 + (sin(2*pi*x[1]))**2)

    def Eggholder(self, x):
        return -(x[0] + 47)*(sin(sqrt(abs((x[1]/2) + (x[0] + 47))))) - (x[1])*(sin(sqrt(abs(x[1] - (x[0] + 47))))) + 959.6407

    def Holder_table(self, x):
        return -(abs((sin(x[0])*(cos(x[1]))*(exp(abs(1 - ((sqrt(x[0]**2 + x[1]**2))/pi))))))) + 19.2085

    def Rastrigin(self, x):
        return (x[0]**2 - 10*cos(2*pi*x[0]) + 10) + (x[1]**2 - 10*cos(2*pi*x[1]) + 10)

    def Schwefel(self, x):
        return 418.9829*(2) - ((x[0]*sin(sqrt(abs(x[0])))) + (x[1]*sin(sqrt(abs(x[1])))))

    def all_func(self):
        return np.array([
            ["Ackley", self.Ackley, self.bound_Ackley],
            ["Levi_N13", self.Levi_N13, self.bound_Levi_N13],
            ["Eggholder", self.Eggholder, self.bound_Eggholder],
            ["Holder_table", self.Holder_table, self.bound_Holder_table],
            ["Rastrigin", self.Rastrigin, self.bound_Rastrigin],
            ["Schwefel", self.Schwefel, self.bound_Schwefel]
            ])

if __name__ == "__main__":
    func = func_2D()
    print(func.n_func)

File: PSO_SHMPSO/Source_code_MATLAB/TestFunction.m
function fitness_error = TestFunction(x, FuncNo)


[m, n] = size(x);

switch FuncNo
    
    % Ackley 1 Function
    case 1
        ninverse = 1 / n;
        sum1 = sum(x .^ 2, 2);
        sum2 = sum(cos(2 * pi * x), 2);
        scores = 20 + exp(1) - (20 * exp(-0.2 * sqrt( ninverse * sum1))) - exp( ninverse * sum2);

    case 2
        scores = (sin(3*pi*x(:, 1)))^2 + ((x(:, 1) - 1)^2)*(1 + (sin(3*pi*x(:, 2)))^2) + ((x(:, 2) - 1)^2)*(1 + (sin(2*pi*x(:, 2)))^2);

    case 3
        scores = -(x(:, 1) + 47)*(sin(sqrt(abs((x(:, 2)/2) + (x(:, 1) + 47))))) - (x(:, 2))*(sin(sqrt(abs(x(:, 2) - (x(:, 1) + 47))))) + 959.6407;

    case 4
        scores = -(abs((sin(x(:, 1))*(cos(x(:, 2)))*(exp(abs(1 - ((sqrt(x(:, 1)^2 + x(:, 2)^2))/pi))))))) + 19.2085;
    % Rastrigin Function
    case 5
        A = 10;
        scores = (A * n) + (sum(x .^2 - A * cos(2 * pi * x), 2));

    case 6
        sum1 = sum((x .*sin(sqrt(abs(x)))), 2);
        scores = 418.9829*n + sum1;

          
end
% switch end

% ่ทๅ–ๆต่ฏ•ๅฝๆ•ฐๆ?ผๅ€?
[~, ~, opt_f] = Get_Func_Info(n, FuncNo);

% ่ฎก็ฎ—่ฏฏๅทฎๅ€?
fitness_error = abs(scores - opt_f);

end